concurrent_events tests: wrong breakpoint hit with multiple concurrent breakpoints

|  |  |
| --- | --- |
| Bugzilla Link | [49433](https://llvm.org/bz49433) |
| Version | unspecified |
| OS | Linux |
| Reporter | LLVM Bugzilla Contributor |
| CC | @JDevlieghere,@labath |

## Extended Description 
All of the tests involving multiple breakpoints suffer from randomly establishing the wrong breakpoint hit count, e.g.:

FAIL: LLDB (/home/mgorny/llvm-project/build.arm64/bin/clang-aarch64) :: test (TestConcurrentTwoBreakpointThreads.ConcurrentTwoBreakpointThreads)
======================================================================
FAIL: test (TestConcurrentTwoBreakpointThreads.ConcurrentTwoBreakpointThreads)
   Test two threads that trigger a breakpoint.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mgorny/llvm-project/lldb/packages/Python/lldbsuite/test/decorators.py", line 149, in wrapper
    return func(*args, **kwargs)
  File "/home/mgorny/llvm-project/lldb/test/API/functionalities/thread/concurrent_events/TestConcurrentTwoBreakpointThreads.py", line 19, in test
    self.do_thread_actions(num_breakpoint_threads=2)
  File "/home/mgorny/llvm-project/lldb/packages/Python/lldbsuite/test/concurrent_base.py", line 266, in do_thread_actions
    breakpoint_hit_count))
AssertionError: 2 != 3 : Expected 2 breakpoint hits, but got 3
Config=aarch64-/home/mgorny/llvm-project/build.arm64/bin/clang


It seems to be a problem in lldb/lldb-server logic somewhere that needs further debugging.

--- a/lldb/packages/Python/lldbsuite/test/lldbutil.py
+++ b/lldb/packages/Python/lldbsuite/test/lldbutil.py
@@ -809,7 +809,7 @@ def is_thread_crashed(test, thread):
             thread.GetStopReason() == lldb.eStopReasonException
             and "EXC_BAD_ACCESS" in thread.GetStopDescription(100)
         )
-    elif test.getPlatform() == "linux":
+    elif test.getPlatform() in ["linux", "freebsd"]:
         return (
             thread.GetStopReason() == lldb.eStopReasonSignal
             and thread.GetStopReasonDataAtIndex(0)

--- a/lldb/test/API/functionalities/thread/concurrent_events/TestConcurrentTwoBreakpointsOneSignal.py
+++ b/lldb/test/API/functionalities/thread/concurrent_events/TestConcurrentTwoBreakpointsOneSignal.py
@@ -8,9 +8,6 @@ class ConcurrentTwoBreakpointsOneSignal(ConcurrentEventsBase):
     # Atomic sequences are not supported yet for MIPS in LLDB.
     @skipIf(triple="^mips")
     @expectedFlakeyNetBSD
-    @expectedFailureAll(
-        archs=["aarch64"], oslist=["freebsd"], bugnumber="llvm.org/pr49433"
-    )
     def test(self):
         """Test two threads that trigger a breakpoint and one signal thread."""
         self.build()

